features/telemetry/protocols/bmp/adapter/src/main/java/org/opennms/netmgt/telemetry/protocols/bmp/adapter/BmpTelemetryAdapter.java

```diff
@@ -0,0 +1,172 @@
+import org.opennms.netmgt.telemetry.protocols.collection.AbstractPersistingAdapter;
+import org.opennms.netmgt.telemetry.protocols.collection.CollectionSetWithAgent;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.codahale.metrics.MetricRegistry;
+
+public class BmpTelemetryAdapter extends AbstractPersistingAdapter {
+
+    private static final Logger LOG = LoggerFactory.getLogger(BmpTelemetryAdapter.class);
+
+    private CollectionAgentFactory collectionAgentFactory;
+
+    private InterfaceToNodeCache interfaceToNodeCache;
+
+    public BmpTelemetryAdapter(final String name,
+                               final MetricRegistry metricRegistry) {
+        super(name, metricRegistry);
+    }
+
+    @Override
+    public Stream<CollectionSetWithAgent> handleMessage(final TelemetryMessageLogEntry message,
+                                                        final TelemetryMessageLog messageLog) {
+        LOG.debug("Received {} telemetry messages", messageLog.getMessageList().size());
+
+        LOG.trace("Parsing packet: {}", message);
+        final BsonDocument document = new RawBsonDocument(message.getByteArray());
+        if (document == null) {
+            return Stream.empty();
+        }
+
+        // This adapter only cares about statistic packets
+        if (!getString(document, "@type")
+                .map(type -> type.equals("STATISTICS_REPORT"))
+                .orElse(false)) {
+            return Stream.empty();
+        }
+
+        // Find the node for the router who has exported the stats and build a collection agent for it
+        final InetAddress exporterAddress = InetAddressUtils.getInetAddress(messageLog.getSourceAddress());
+        final Optional<Integer> exporterNodeId = this.interfaceToNodeCache.getFirstNodeId(messageLog.getLocation(), exporterAddress);
+        if (!exporterNodeId.isPresent()) {
+            LOG.warn("Unable to find node and interface for agent address: {}", exporterAddress);
+            return Stream.empty();
+        }
+        final CollectionAgent agent = this.collectionAgentFactory.createCollectionAgent(Integer.toString(exporterNodeId.get()), exporterAddress);
+
+        // Extract peer details
+        final String peerAddress = getString(document, "peer", "address").get();
+        final Instant timestamp = Instant.ofEpochSecond(getInt64(document, "peer", "timestamp", "epoch").get(),
+                                                        getInt64(document, "peer", "timestamp", "nanos").orElse(0L));
+        final String as = Long.toString(getInt64(document, "peer", "as").get());
+        final String id = Long.toString(getInt64(document, "peer", "id").get());
+
+        // Build resource for the peer
+        final NodeLevelResource nodeResource = new NodeLevelResource(agent.getNodeId());
+        final GenericTypeResource peerResource = new GenericTypeResource(nodeResource, RESOURCE_TYPE, peerAddress);
```

Comment:
Can use a `DeferredGenericTypeResource` instead so we don't need to define it in code. The resource type should be added to the config in `resource-types.d`.

Reply:
Didn't know about that. Changed.